### ISSUE TITLE
piper tts, text -> phoenemese

### ISSUE THREAD

**Issue by m172021** [robertknight/rten#194]
Is there a way to replace https://github.com/robertknight/rten/blob/main/rten-examples/src/piper.rs#L187-L193 with http://www.speech.cs.cmu.edu/cgi-bin/cmudict/ (or some other tool that runs in Rust) ?

Is the piper 'phoeneme' dictionary a standard dictionary (i.e. we can use other tools for text -> phoeneme) or a piper specific format ?

Thanks!

**Comment by robertknight** [comment]
Piper uses [piper-phonemize](https://github.com/rhasspy/piper-phonemize), which is a C++ wrapper around a [modified version](https://github.com/rhasspy/espeak-ng) of espeak-ng. The closest thing to a drop-in replacement for Rust would probably be something like the [espeakng](https://crates.io/crates/espeakng) crate.

The Piper author mentioned in [this discussion](https://github.com/rhasspy/piper/discussions/504) that they are planning to make Piper's text-to-phoneme library a standalone C++ lib in future, which will make it easier to wrap in Rust.

**Comment by robertknight** [comment]
> Is the piper 'phoeneme' dictionary a standard dictionary (i.e. we can use other tools for text -> phoeneme) or a piper specific format ?

The phonemes use the IPA alphabet, so you can likely get _something_ out if you use an alternative text -> IPA tool. Whether it will be close enough to Piper's own translation to get _good_ output is another matter.

**Comment by m172021** [comment]
Clearly I am misunderstanding something, and would appreciate your insights on where I'm going wrong.

1. Let us define EnglishWord to be a finite subset of String.
2. My understanding is that there is a 'standard' mapping of `HashMap<EnglishWord, Vec<IPA>>` -- for example, at https://svn.code.sf.net/p/cmusphinx/code/trunk/cmudict/cmudict-0.7b

A few concrete examples of above:

UNILATERALLY  Y UW2 N AH0 L AE1 T ER0 AH0 L IY0
UNILATERALLY(1)  Y UW2 N AH0 L AE1 T R AH0 L IY0
UNILEVER  Y UW2 N IH0 L IY1 V ER0
UNILEVER'S  Y UW2 N IH0 L IY1 V ER0 Z
UNIMAGINABLE  AH2 N IH0 M AE1 JH IH0 N AH0 B AH0 L
UNIMAGINATIVE  AH0 N IH0 M AE1 JH IH2 N AH0 T IH0 V
UNIMAGINED  AH2 N IH0 M AE1 JH AH0 N D

Note, the word 'Unilaterally' seems to have 2 "standard" conversions.

So my idea was something like:

1. load that cmu dict as a `HashMap<String, Vec<IPA>>`

2. For "string -> wav", do: split into words, index into that hashmap, get the `Vec<IPA>`, feed that to rten/piper

Clearly you have thought about this far more than I have -- and I would love to hear your 'criticism' (looking for brutal honesty, not superficial kindness) on the flaws of the above logic.

**Comment by robertknight** [comment]
Your idea will partially work, but note the original CMUdict doesn't use IPA. Instead you'll need to find an IPA conversion of the dictionary such as https://github.com/menelik3/cmudict-ipa. If you take the IPA example strings from that repo's README and feed them into the demo, they produce reasonable output:

```
cargo run -p rten-examples -r --bin piper voice.rten voice.onnx.json "ˌhɪməˈleɪən, ˌhɪməˈlaɪən"
```

However, if you try other words in this version of CMUdict, you'll find that many words don't work as well. You can play with Piper's own translation using Python:

```
pip install piper-phonemize
python

>>> import piper_phonemize as pp
>>> output = pp.phonemize_espeak("unilateral", "en-us")
>>> "".join(output[0])
'jˌuːnɪlˈæɾɚɹəl'
```

What you could do is write a script that takes a large list of words and uses this Python module to generate a word-to-phoneme dictionary, in a format you can load into Rust. Then you can translate a sentence using a series of dictionary lookups.

**Comment by m172021** [comment]
> What you could do is write a script that takes a large list of words and uses this Python module to generate a word-to-phoneme dictionary, in a format you can load into Rust. Then you can translate a sentence using a series of dictionary lookups.

This is clever. Just brute force the English dictionary.

What is the weakness of this idea? Why has it not been already done? (Only weakness I see is trying to sound out English-looking but non-English words.)

**Comment by m172021** [comment]
Closing the issue as @robertknight  is a %*#($*# genius coder and I have nothing else to add.

**Comment by robertknight** [comment]
> What is the weakness of this idea? Why has it not been already done? (Only weakness I see is trying to sound out English-looking but non-English words.)

As obvious as it seems to me now, I just didn't think of it when I created the demo last week.

I started working on a proof of concept in https://github.com/robertknight/rten/pull/197. It works, as long as your input consists entirely of words that appear in the dictionary.

The main weakness is that it is incomplete - it doesn't handle:

- Words that don't appear in the dictionary
- Numbers
- Punctuation
- Ways in which some words might affect the pronunciation of others in the sentence (eg. homophones)
- A whole bunch of other things I'm not aware of, given my limited knowledge of this domain

I will get the draft landed, but fleshing it out fully to with all of the above is outside the scope of the example.

Another different direction that occurred to me would be to treat text-to-phoneme conversion as a language translation task and train a simple sequence-to-sequence model on a bunch of example text.
